python_modules/dagster/dagster/_core/ext/context.py

```diff
@@ -19,34 +22,84 @@
 from dagster._core.definitions.events import AssetKey
 from dagster._core.definitions.metadata import MetadataValue, normalize_metadata_value
 from dagster._core.definitions.partition_key_range import PartitionKeyRange
+from dagster._core.definitions.result import MaterializeResult
 from dagster._core.definitions.time_window_partitions import TimeWindow
 from dagster._core.execution.context.compute import OpExecutionContext
 from dagster._core.execution.context.invocation import BoundOpExecutionContext
+from dagster._core.ext.client import ExtMessageReader
 
 
 class ExtMessageHandler:
     def __init__(self, context: OpExecutionContext) -> None:
         self._context = context
+        self._asset_materialized_map: Dict[AssetKey, bool] = {
+            key: False for key in context.selected_asset_keys
+        }
+        self._result_queue: List[MaterializeResult] = []
+        self._metadata: Dict[AssetKey, Dict[str, MetadataValue]] = {}
+        self._data_versions: Dict[AssetKey, DataVersion] = {}
+
+        # The lock is used to guard modification of the asset materialization tracking data
+        # structures, which are modified by both handler methods (and thus the message reader
+        # thread) and the `clear_materialization_queue` method, which can be called by the
+        # main thread.
+        self._lock = Lock()
+
+    @contextmanager
+    def handle_messages(self, message_reader: ExtMessageReader) -> Iterator[ExtParams]:
+        with message_reader.read_messages(self) as params:
+            yield params
+        # enqueue any remaining assets
+        unmaterialized_assets = {k for k, v in self._asset_materialized_map.items() if not v}
+        for key in unmaterialized_assets:
+            self._result_queue.append(self._materialize_result_for_asset(key))
+            self._asset_materialized_map[key] = True
+
+    def clear_result_queue(self) -> Iterator[MaterializeResult]:
+        with self._lock:
+            while self._result_queue:
+                yield self._result_queue.pop(0)
+
+    def _materialize_result_for_asset(self, asset_key: AssetKey) -> MaterializeResult:
+        x = MaterializeResult(
+            asset_key=asset_key,
+            metadata=self._metadata.get(asset_key),
+            data_version=self._data_versions.get(asset_key),
+        )
+        return x
 
     # Type ignores because we currently validate in individual handlers
     def handle_message(self, message: ExtMessage) -> None:
-        if message["method"] == "report_asset_metadata":
-            self._handle_report_asset_metadata(**message["params"])  # type: ignore
-        elif message["method"] == "report_asset_data_version":
-            self._handle_report_asset_data_version(**message["params"])  # type: ignore
-        elif message["method"] == "log":
-            self._handle_log(**message["params"])  # type: ignore
+        with self._lock:
```

Comment:
I don't understand we aren't pushing down this lock closer to the operation. We should acquire the lock _only_ for mutating the queue, not for extraneous operations. Calling handle._log is particularly egregious here. Let's wrap the queue and create a lightweight `LockingQueue` abstraction to avoid this mistake.

Reply:
Good point, I've now removed the `Lock` entirely in favor of `queue.Queue` which is already thread-safe.